src/dragonball/src/dbs_pci/src/msi.rs

```diff
@@ -0,0 +1,759 @@
+
+        if let Err(e) = self.read(offset as u64, &mut buf) {
+            debug!("failed to read PCI MSI capability structure, {}", e);
+            fill_config_data(&mut buf);
+        }
+
+        buf[0]
+    }
+
+    fn read_u16(&mut self, offset: usize) -> u16 {
+        let mut buf = [0u8; 2];
+
+        if let Err(e) = self.read(offset as u64, &mut buf) {
+            debug!("failed to read PCI MSI capability structure, {}", e);
+            fill_config_data(&mut buf);
+        }
+
+        LittleEndian::read_u16(&buf)
+    }
+
+    fn read_u32(&mut self, offset: usize) -> u32 {
+        let mut buf = [0u8; 4];
+
+        if let Err(e) = self.read(offset as u64, &mut buf) {
+            debug!("failed to read PCI MSI capability structure, {}", e);
+            fill_config_data(&mut buf);
+        }
+
+        LittleEndian::read_u32(&buf)
+    }
+
+    fn write_u8(&mut self, offset: usize, value: u8) {
+        if let Err(e) = self.write(offset as u64, &[value]) {
+            debug!("failed to write PCI MSI capability structure, {}", e);
+        }
+    }
+
+    fn write_u16(&mut self, offset: usize, value: u16) {
+        let mut buf = [0u8; 2];
+        LittleEndian::write_u16(&mut buf, value);
+        if let Err(e) = self.write(offset as u64, &buf) {
+            debug!("failed to write PCI MSI capability structure, {}", e);
+        }
+    }
+
+    fn write_u32(&mut self, offset: usize, value: u32) {
+        let mut buf = [0u8; 4];
+        LittleEndian::write_u32(&mut buf, value);
+        if let Err(e) = self.write(offset as u64, &buf) {
+            debug!("failed to write PCI MSI capability structure, {}", e);
+        }
+    }
+
+    fn pci_capability_type(&self) -> PciCapabilityID {
+        PciCapabilityID::MessageSignalledInterrupts
+    }
+}
+
+/// Struct to manage PCI Message Signalled Interrupt controller working state.
+#[repr(packed)]
+#[derive(Clone, Copy, Default, PartialEq)]
+pub struct MsiState {
+    msg_ctl: u16,
+    msg_addr_lo: u32,
+    msg_addr_hi: u32,
+    msg_data: u16,
+    mask_bits: u32,
+}
+
+impl MsiState {
+    /// Create a new PCI MSI capability state structure.
+    pub fn new(mut msg_ctl: u16) -> Self {
+        // Initialize as disabled
+        msg_ctl &= !MSI_CTL_ENABLE;
+
+        MsiState {
+            msg_ctl,
+            ..Default::default()
+        }
+    }
+
+    /// Check whether the PCI MSI capability has been enabled.
+    pub fn enabled(&self) -> bool {
+        self.msg_ctl & MSI_CTL_ENABLE == MSI_CTL_ENABLE
+    }
+
+    /// Get number of vectors supported.
+    pub fn num_vectors(&self) -> u16 {
+        let field = (self.msg_ctl >> 1) & 0x7;
+
+        if field > 5 {
+            return 0;
+        }
+
+        1 << field
+    }
+
+    /// Get number of vectors enabled.
+    pub fn num_enabled_vectors(&self) -> u16 {
+        let field = (self.msg_ctl >> 4) & 0x7;
+
+        if field > 5 {
+            return 0;
+        }
+
+        1 << field
+    }
```

Comment:
Duplicated code.

Reply:
this method will call `msi_num_enabled_vectors` to reduce duplication